pkg/handler/handler.go

```diff
@@ -96,26 +98,78 @@ type ReadmeUtgaveFilter struct {
 	First  *int32
 }
 
-func (r *resolver) ReadmeUtgaver(filter *ReadmeUtgaveFilter) ([]ReadmeUtgave, error) {
-	if filter == nil {
-		return r.readmes, nil
-	}
-	var filteredReadmes []ReadmeUtgave
-	for _, r := range r.readmes {
-		if filter.Year != nil && *filter.Year != r.Year {
-			continue
+func getReadmes(ctx context.Context, client storage.Client, query string) ([]ReadmeUtgave, error) {
+	pdfQuery := &storage.Query{Prefix: "pdf/" + query}
+	imageQuery := &storage.Query{Prefix: "images/" + query}
+
+	re := regexp.MustCompile(`(?P<Year>\d{4})-(?P<Utgave>\d{2})`)
+
+	bkt := client.Bucket("readme-arkiv.appspot.com")
+	pdfIt := bkt.Objects(ctx, pdfQuery)
+	imageIt := bkt.Objects(ctx, imageQuery)
+	var err error
+	var readmes  []ReadmeUtgave
+	for {
+		pdfAttrs, err := pdfIt.Next()
+		if err == iterator.Done {
+			break
 		}
-		if filter.Utgave != nil && *filter.Utgave != r.Utgave {
+		imageAttrs, err := imageIt.Next()
+
+		if err == iterator.Done {
+			break
+		}
+
+		// Object may be a directory
+		match := re.FindAllString(pdfAttrs.Name, -1)
+		if len(match) == 0 {
 			continue
 		}
-		filteredReadmes = append(filteredReadmes, r)
-		if filter.First != nil && len(filteredReadmes) == int(*filter.First) {
-			break
+
+		matches := re.FindStringSubmatch(pdfAttrs.Name)
+		utgave, _ := strconv.ParseInt(matches[2], 10, 32)
+		year, _ := strconv.ParseInt(matches[1], 10, 32)
+		title := re.FindAllString(pdfAttrs.Name, 1)[0]
+
+		r := ReadmeUtgave{
+			Title: title,
+			Image: imageAttrs.MediaLink,
```

Comment:
The bucket contents are not public by default, so the media link is not a public URL. 

Solutions could be to use a signed URL[1] or make the contents of the bucket public. The former is more "proper", but the latter is definitively easier. Do you have a preference?

[1] [https://godoc.org/cloud.google.com/go/storage#SignedURL](https://godoc.org/cloud.google.com/go/storage#SignedURL)

Reply:
I was going to mention this, and I think the best way is to make the relevant directories public, as the signed urls will expire after some time, and expiring urls are definitely not the best. As far as I see, there shouldn't really be a reason not to have the bucket(directories if possible) public?

Reply:
I've had a poke around in the settings, and it seems like its an all or nothing game. I can't make individual folders public, it seems. I can, however, set individual objects to public.

I could set the objects to public individually on upload and write a single-use Cloud Function to make the objects that are already there public. This would allow us to expand the service with other types of data that should not be public in the future, but for now, the simplest solution is by far to make the bucket public.

It should be open now. I'll probably get e-mails from Google nagging me about this 😆

Reply:
Cool :+1:. I think the proper way to do this is just to have a separate bucket for public data, and just keep private data on another bucket. But if you want to mix the two, have a look at https://cloud.google.com/iam/docs/conditions-attribute-reference#resourcename_attribute